src/MultiFormats/Registry/MultiBaseAlgorithm .cs

```diff
@@ -0,0 +1,224 @@
+        public string Name { get; private set; }
+
+        /// <summary>
+        ///   The IPFS code assigned to the algorithm.
+        /// </summary>
+        /// <value>
+        ///   Valid codes at <see href="https://github.com/multiformats/multibase/blob/master/multibase.csv"/>.
+        /// </value>
+        public char Code { get; private set; }
+
+        /// <summary>
+        ///   Returns a function that can return a string from a byte array.
+        /// </summary>
+        public Func<byte[], string> Encode { get; private set; }
+
+        /// <summary>
+        ///   Returns a function that can return a byte array from a string.
+        /// </summary>
+        public Func<string, byte[]> Decode { get; private set; }
+
+        /// <summary>
+        ///   Use <see cref="Register"/> to create a new instance of a <see cref="MultiBaseAlgorithm"/>.
+        /// </summary>
+        private MultiBaseAlgorithm() { }
+
+        /// <summary>
+        ///   The <see cref="Name"/> of the algorithm.
+        /// </summary>
+        public override string ToString() { return Name; }
+
+        /// <summary>
+        ///   Register a new IPFS algorithm.
+        /// </summary>
+        /// <param name="name">
+        ///   The name of the algorithm.
+        /// </param>
+        /// <param name="code">
+        ///   The IPFS code assigned to thealgorithm.
+        /// </param>
+        /// <param name="encode">
+        ///   A <c>Func</c> to encode a byte array.  If not specified, then a <c>Func</c> is created to
+        ///   throw a <see cref="NotImplementedException"/>.
+        /// </param>
+        /// <param name="decode">
+        ///   A <c>Func</c> to decode a string.  If not specified, then a <c>Func</c> is created to
+        ///   throw a <see cref="NotImplementedException"/>.
+        /// </param>
+        /// <returns>
+        ///   A new <see cref="MultiBaseAlgorithm"/>.
+        /// </returns>
+        public static MultiBaseAlgorithm Register(string name,
+            char code,
+            Func<byte[], string> encode = null,
+            Func<string, byte[]> decode = null)
+        {
+            if (string.IsNullOrWhiteSpace(name))
+                throw new ArgumentNullException("name");
+            if (Names.ContainsKey(name))
+                throw new ArgumentException(
+                    string.Format("The IPFS multi-base algorithm name '{0}' is already defined.", name));
+            if (Codes.ContainsKey(code))
+                throw new ArgumentException(
+                    string.Format("The IPFS multi-base algorithm code '{0}' is already defined.", code));
+            if (encode == null)
+                encode = (bytes) =>
+                {
+                    throw new NotImplementedException(
+                        string.Format("The IPFS encode multi-base algorithm '{0}' is not implemented.", name));
+                };
+
+            if (decode == null)
+                decode = (s) =>
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Introduce a new variable instead of reusing the parameter 'decode'.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)